edit-post/components/plugin-more-menu-item/index.js

```diff
@@ -0,0 +1,83 @@
+/**
+ * External dependencies
+ */
+import { noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { compose } from '@wordpress/element';
+import { Slot, Fill, withContext, MenuItemsGroup, MenuItemsItem } from '@wordpress/components';
+import { withDispatch, withSelect } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Name of slot in which the sidebar should fill.
+ *
+ * @type {String}
+ */
+const SLOT_NAME = 'PluginMoreMenuItem';
+
+function PluginMoreMenuItem( { label, onClick, icon, isSelected } ) {
+	return (
+		<Fill name={ SLOT_NAME }>
+			{ ( props ) => {
+				return (
+					<MenuItemsItem
+						icon={ isSelected ? 'yes' : icon }
+						isSelected={ isSelected }
+						label={ label }
+						onClick={ () => {
+							onClick();
+							props.onClose();
+						} }
+					/>
+				);
+			} }
+		</Fill>
+	);
+}
+
+PluginMoreMenuItem = compose( [
+	withContext( 'pluginName' )( ( pluginName, { target } ) => {
+		return {
+			target: `${ pluginName }/${ target }`,
+		};
+	} ),
+	withSelect( ( select, { target } ) => ( {
+		isSelected: select( 'core/edit-post' ).getActiveGeneralSidebarName() === target,
+	} ) ),
+	withDispatch( ( dispatch, { type, target, isSelected } ) => {
+		let onClick = noop;
+		if ( isSelected ) {
+			onClick = dispatch( 'core/edit-post' ).closeGeneralSidebar;
+		} else {
+			switch ( type ) {
+				case 'sidebar':
+					onClick = () => dispatch( 'core/edit-post' ).openGeneralSidebar( target );
```

Comment:
Now, on the other hand, some nice destructuring that can help readability is to destructure dispatching action creators at the top of the function:

```js
const { closeGeneralSidebar, openGeneralSidebar } = dispatch( 'core/edit-post' );

// ...

onClick = () => openGeneralSidebar( target );
```